base_layer/core/src/base_node/states/horizon_state_sync/state_sync.rs

```diff
@@ -0,0 +1,705 @@
+                if attempt == config.max_sync_request_retry_attempts {
+                    return Err(HorizonSyncError::MaxSyncAttemptsReached);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    // Finalize the horizon state synchronization by setting the chain metadata to the local tip and committing the
+    // horizon state to the blockchain backend.
+    async fn finalize_horizon_sync(&self) -> Result<(), HorizonSyncError> {
+        // TODO: Perform final validation on full synced horizon state before committing horizon state checkpoint.
+        // TODO: Verify header and kernel set using Mmr Roots
+
+        async_db::commit_horizon_state(self.db()).await?;
+
+        Ok(())
+    }
+
+    // Validate the received UTXO set and, UTXO and RangeProofs MMR nodes.
+    fn validate_utxos_and_rangeproofs(
+        &self,
+        utxo_hashes: &[HashOutput],
+        rp_hashes: &[HashOutput],
+        request_utxo_hashes: &[HashOutput],
+        request_rp_hashes: &[HashOutput],
+        utxos: &[TransactionOutput],
+    ) -> Result<(), HorizonSyncError>
+    {
+        if utxo_hashes.is_empty() | rp_hashes.is_empty() | utxos.is_empty() {
+            return Err(HorizonSyncError::EmptyResponse);
+        }
+        // Check if the same number of utxo and rp MMR nodes returned
+        if utxo_hashes.len() != rp_hashes.len() {
+            return Err(HorizonSyncError::IncorrectResponse);
+        }
+        // Check that the correct number of utxos returned
+        if request_utxo_hashes.len() != utxos.len() {
+            return Err(HorizonSyncError::IncorrectResponse);
+        }
+
+        // Check that utxo set is the requested utxos
+        if (0..request_utxo_hashes.len()).any(|i| utxos[i].hash() != request_utxo_hashes[i]) {
+            return Err(HorizonSyncError::IncorrectResponse);
+        }
+
+        // Check that utxo set matches the provided RangeProof MMR Nodes
+        if (0..request_rp_hashes.len()).any(|i| utxos[i].proof.hash() != request_rp_hashes[i]) {
+            return Err(HorizonSyncError::IncorrectResponse);
+        }
+
+        Ok(())
+    }
+
+    // Validate the received UTXO set and, UTXO and RangeProofs MMR nodes.
+    fn validate_utxo_hashes(
+        &self,
+        local_utxo_hashes: &[HashOutput],
+        remote_utxo_hashes: &[HashOutput],
+    ) -> Result<(), HorizonSyncError>
+    {
+        // Check that the correct number of utxo hashes returned
+        if local_utxo_hashes.len() != remote_utxo_hashes.len() {
+            return Err(HorizonSyncError::IncorrectResponse);
+        }
+
+        // Check that the received utxo set is the same as local
+        if (0..local_utxo_hashes.len()).any(|i| local_utxo_hashes[i] != remote_utxo_hashes[i]) {
+            return Err(HorizonSyncError::IncorrectResponse);
+        }
+
+        Ok(())
+    }
+
+    // Check the received set of kernels.
+    fn validate_kernels(
```

Comment:
we should validate the kernel signatures as well.

Reply:
Agreed, I have added that in another branch that I want to PR that validates the final state